## Import Libraries

In [2]:
import pandas as pd


StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 4, Finished, Available, Finished, False)

## Load Dataset 

In [3]:
path = "/lakehouse/default/Files/bronze_superstore_sales.csv"

data = pd.read_csv(path, encoding="latin-1")

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 5, Finished, Available, Finished, False)

### Data Inspection

In [4]:
data.head()

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 6, Finished, Available, Finished, False)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [5]:
data.columns.to_list()

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 7, Finished, Available, Finished, False)

['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit']

#### Data Inspection

The dataset is a standard retail order table — part of a broader end-to-end sales analysis pipeline. Each row represents one order line.

#### Columns in Dataset

Identifiers (Row ID, Order ID, Customer ID, Product ID) link each row to a specific transaction, customer, and product. Order Date, Ship Date, and Ship Mode cover the time and logistics side. Location goes down to postal code: Country, City, State, Postal Code, and Region. Customer fields include Customer Name and Segment. Product fields run three levels deep: Category → Sub-Category → Product Name. The financial columns are Sales, Quantity, Discount, and Profit.



## Dataset Information

In [6]:
data.shape

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 8, Finished, Available, Finished, False)

(9994, 21)

In [7]:
data.info()

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 9, Finished, Available, Finished, False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [8]:
data.isnull().sum()

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 10, Finished, Available, Finished, False)

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

In [9]:
data.duplicated().sum()

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 11, Finished, Available, Finished, False)

0

#### Dataset Information

The dataset has **9,994 rows** and **21 columns** — enough depth to run meaningful segmentation, trend, and profitability analysis without sampling.

#### Data Quality Checks

No null values. No duplicate records. The dataset is clean on both counts.

The one thing that needs fixing before analysis: data types. A few columns — mainly the date fields — aren't in the right format yet and will be corrected in the preprocessing step.

#### What This Means for the Transformation

Most of the cleaning work here is minimal. The only real task is type-casting the date columns to proper datetime format so time-based calculations work correctly downstream.

## Data Transformation

#### Converting Object to Datetime Datatype

In [10]:
data['Order Date'] = pd.to_datetime(data['Order Date'])
data['Ship Date'] = pd.to_datetime(data['Ship Date'])

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 12, Finished, Available, Finished, False)

#### Converting Categorical Columns Datatype AS Category Datatype

In [11]:
categorical_cols = [
'Ship Mode',
'Segment',
'Country',
'City',
'State',
'Region',
'Category',
'Sub-Category'
]

for col in categorical_cols:
    data[col] = data[col].astype('category')

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 13, Finished, Available, Finished, False)

#### Converting Keys Columns Datatype AS String Datatype

In [12]:
id_cols = [
'Order ID',
'Customer ID',
'Product ID',
'Postal Code'
]

for col in id_cols:
    data[col] = data[col].astype('string')

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 14, Finished, Available, Finished, False)

#### Renaming column's names to snake case

In [13]:
data.columns = (
    data.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
    .str.replace("-", "_")
)

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 15, Finished, Available, Finished, False)

#### Data Transformation

Four things were done to get the dataset analysis-ready: date parsing, categorical encoding, identifier typing, and column renaming.

#### Transformations Applied

**Datetime Conversion**  
`Order Date` and `Ship Date` were converted from object to datetime. Without this, any time-based calculation — trends, delivery gaps, seasonality — won't work correctly.

**Categorical Conversion**  
Ship Mode, Segment, Country, City, State, Region, Category, and Sub-Category were converted to category dtype. These columns have a small number of repeating values, so storing them as categories cuts memory usage and speeds up groupby operations noticeably.

**String Conversion for Identifiers**  
Order ID, Customer ID, Product ID, and Postal Code were cast to string. They're labels, not numbers — keeping them as objects or integers risks unintended arithmetic and inconsistent joins downstream.

**Column Renaming**  
All column names were renamed to snake_case. Consistent naming removes the friction of quoting columns with spaces in Pandas, SQL, and Power BI.

#### Result

The dataset is now typed correctly, lighter in memory, and consistent in naming — ready to load into any analysis or visualization layer without further prep.


## Feature Engineering

#### Creating Date Features Like Year & Month

In [14]:
data['order_year'] = data['order_date'].dt.year
data['order_month'] = data['order_date'].dt.month

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 16, Finished, Available, Finished, False)

#### Creating Shipping Days Column

In [15]:
data['shipping_days'] = (data['ship_date'] - data['order_date']).dt.days

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 17, Finished, Available, Finished, False)

#### Creating Profit Margin & Percentage Columns

In [16]:
data['profit_margin'] = data['profit'] / data['sales']

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 18, Finished, Available, Finished, False)

In [17]:
data['profit_margin_pct'] = (data['profit'] / data['sales']) * 100

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 19, Finished, Available, Finished, False)

#### Feature Engineering

Four new columns were added to make the data more directly useful for business analysis. None of these existed in the raw dataset — they're all derived from existing fields.

#### Features Created

**Date Features**  
Year and Month were extracted from `order_date`. Having these as standalone columns makes time-based SQL and Power BI straightforward — no need to parse dates repeatedly during analysis.

**Shipping Duration**  
`shipping_days` was created by subtracting `order_date` from `ship_date`. This gives a concrete measure of delivery speed per order, useful for spotting logistics bottlenecks by region, ship mode, or time period.

**Profit Metrics**  
Two profitability columns were added: `profit_margin` (profit relative to sales) and `profit_percentage`. These make it easier to compare performance across products, categories, and regions without doing the calculation every time in a query or visual.

#### Result

The dataset now has the time, logistics, and profitability dimensions pre-built. Downstream analysis can work directly with these columns instead of re-deriving them at every step.


## Final Validation

In [18]:
data.info()

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 20, Finished, Available, Finished, False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   row_id             9994 non-null   int64         
 1   order_id           9994 non-null   string        
 2   order_date         9994 non-null   datetime64[ns]
 3   ship_date          9994 non-null   datetime64[ns]
 4   ship_mode          9994 non-null   category      
 5   customer_id        9994 non-null   string        
 6   customer_name      9994 non-null   object        
 7   segment            9994 non-null   category      
 8   country            9994 non-null   category      
 9   city               9994 non-null   category      
 10  state              9994 non-null   category      
 11  postal_code        9994 non-null   string        
 12  region             9994 non-null   category      
 13  product_id         9994 non-null   string        
 14  category

In [19]:
data['order_year'] = data['order_year'].astype('int16')
data['order_month'] = data['order_month'].astype('int8')

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 21, Finished, Available, Finished, False)

In [20]:
data['shipping_days'] = data['shipping_days'].astype('int16')

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 22, Finished, Available, Finished, False)

In [21]:
data['profit_margin'] = data['profit_margin'].astype('float32')

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 23, Finished, Available, Finished, False)

#### Final Validation

The last step was downcasting the newly created columns to smaller integer and float types. The raw dataset only has 9,994 rows, so this isn't about survival — it's just good practice, especially if this pipeline gets reused on larger data later.

#### Data Type Optimization

`order_year` and `shipping_days` were cast to int16. `order_month` went to int8 — month values only ever run 1 to 12, so int64 is overkill. `profit_margin` was cast to float32, which holds enough decimal precision for margin calculations without the extra memory overhead of float64.

None of these changes affect the values or analysis output. They just make the dataset lighter.

#### Result

All columns are now typed to match their actual value ranges. The dataset is clean, consistently named, correctly typed, and ready to export 

## Spark DataFrame Creation & Delta Table Storage

In [22]:
spark_df = spark.createDataFrame(data)

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 24, Finished, Available, Finished, False)

In [23]:
display(spark_df)

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 25, Finished, Available, Finished, False)

SynapseWidget(Synapse.DataFrame, 6edf52e9-4ed7-496a-8ccc-45e71e726214)

In [24]:
spark_df.write.format("delta").mode("overwrite").saveAsTable("silver_sales_cleaned")

StatementMeta(, cc129f69-31b5-4634-b1e7-b93711ca5c10, 26, Finished, Available, Finished, False)

#### Spark DataFrame Creation & Delta Table Storage

At this point the data is clean, typed, and validated. The last step is moving it out of Pandas processing and into a format the rest of the pipeline can actually use at scale.

#### Process Breakdown

**Pandas to Spark**  
The cleaned Pandas DataFrame was converted to a Spark DataFrame using `spark.createDataFrame(data)`. This hands off the data to distributed processing — necessary once the pipeline moves beyond a single notebook or dataset size grows.

**Data Preview**  
`display(spark_df)` was used to do a quick sanity check on the schema and first few records before writing anything to storage. Small step, but worth doing before an overwrite.

**Delta Table Storage**  
The Spark DataFrame was written using `write.format("delta").mode("overwrite").saveAsTable("silver_sales_cleaned")`. Delta Lake handles the storage layer here — it gives ACID transactions, version history (time travel if something breaks), and consistent reads even during writes. Overwrite mode replaces the existing table cleanly without stacking duplicates.

**Silver Layer Placement**  
The table is saved as `silver_sales_cleaned`, which places it in the Silver layer of the medallion architecture: Bronze holds raw ingested data, Silver holds cleaned and transformed data, and Gold holds business-ready aggregations. This notebook covers the Bronze → Silver step.

#### Result

The transformed dataset is now stored as a managed Delta table, queryable from SQL, consumable by downstream notebooks, and ready to feed into the Gold layer
